From 19cd989768c3abbb107491c8c1e69fa3d4a31ece Mon Sep 17 00:00:00 2001
From: Tobias <ruck.tobias@gmail.com>
Date: Tue, 13 Apr 2021 09:33:27 -0500
Subject: [PATCH 076/196] Make Taproot standard (#136)

Allows Taproot outputs and spends to occur in mempool and block transactions.
Doesn't add any json/address/wallet support yet.
---
 src/script/sign.cpp                           |  1 +
 src/script/standard.cpp                       | 15 ++++++
 src/script/standard.h                         |  1 +
 src/test/script_standard_tests.cpp            | 47 +++++++++++++++++++
 src/test/transaction_tests.cpp                | 34 ++++++++++++++
 src/wallet/rpcdump.cpp                        |  1 +
 src/wallet/scriptpubkeyman.cpp                |  1 +
 .../logos_feature_taproot_key_spend.py        |  2 -
 .../logos_feature_taproot_script_path.py      |  2 -
 9 files changed, 100 insertions(+), 4 deletions(-)

diff --git a/src/script/sign.cpp b/src/script/sign.cpp
index 4985980a7..fc3542cfb 100644
--- a/src/script/sign.cpp
+++ b/src/script/sign.cpp
@@ -124,6 +124,7 @@ static bool SignStep(const SigningProvider &provider,
     switch (whichTypeRet) {
         case TxoutType::NONSTANDARD:
         case TxoutType::NULL_DATA:
+        case TxoutType::TAPROOT:
             return false;
         case TxoutType::PUBKEY:
             if (!CreateSig(creator, sigdata, provider, sig,
diff --git a/src/script/standard.cpp b/src/script/standard.cpp
index ac949f73c..5fade1bcc 100644
--- a/src/script/standard.cpp
+++ b/src/script/standard.cpp
@@ -7,6 +7,7 @@
 
 #include <crypto/sha256.h>
 #include <script/script.h>
+#include <script/taproot.h>
 
 #include <string>
 
@@ -41,6 +42,8 @@ std::string GetTxnOutputType(TxoutType t) {
             return "scripthash";
         case TxoutType::MULTISIG:
             return "multisig";
+        case TxoutType::TAPROOT:
+            return "taproot";
         case TxoutType::NULL_DATA:
             return "nulldata";
     } // no default case, so the compiler can warn about missing cases
@@ -122,6 +125,14 @@ TxoutType Solver(const CScript &scriptPubKey,
         return TxoutType::SCRIPTHASH;
     }
 
+    // Taproot output:
+    // OP_SCRIPTTYPE OP_1 33 <33-byte commitment>
+    // or
+    // OP_SCRIPTTYPE OP_1 33 <33-byte commitment> 32 <32-byte state>
+    if (IsPayToTaproot(scriptPubKey)) {
+        return TxoutType::TAPROOT;
+    }
+
     // Provably prunable, data-carrying output
     //
     // So long as script passes the IsUnspendable() test and all but the first
@@ -180,6 +191,10 @@ bool ExtractDestination(const CScript &scriptPubKey,
         addressRet = ScriptHash(uint160(vSolutions[0]));
         return true;
     }
+    if (whichType == TxoutType::TAPROOT) {
+        // Taproot doesn't has destinations for now
+        return false;
+    }
     // Multisig txns have more than one address...
     return false;
 }
diff --git a/src/script/standard.h b/src/script/standard.h
index 7d606b9e3..5ad21a9a6 100644
--- a/src/script/standard.h
+++ b/src/script/standard.h
@@ -89,6 +89,7 @@ enum class TxoutType {
     PUBKEYHASH,
     SCRIPTHASH,
     MULTISIG,
+    TAPROOT,
     // unspendable OP_RETURN script that carries data
     NULL_DATA,
 };
diff --git a/src/test/script_standard_tests.cpp b/src/test/script_standard_tests.cpp
index 12df1645f..c76cc6fba 100644
--- a/src/test/script_standard_tests.cpp
+++ b/src/test/script_standard_tests.cpp
@@ -97,6 +97,18 @@ BOOST_AUTO_TEST_CASE(script_standard_Solver_success) {
     BOOST_CHECK(solutions[3] == ToByteVector(pubkeys[2]));
     BOOST_CHECK(solutions[4] == std::vector<uint8_t>({3}));
 
+    // TxoutType::TAPROOT
+    s.clear();
+    s << OP_SCRIPTTYPE << OP_1 << ToByteVector(pubkeys[0]);
+    BOOST_CHECK_EQUAL(Solver(s, solutions), TxoutType::TAPROOT);
+    BOOST_CHECK_EQUAL(solutions.size(), 0U);
+
+    s.clear();
+    s << OP_SCRIPTTYPE << OP_1 << ToByteVector(pubkeys[0])
+      << std::vector<uint8_t>(32);
+    BOOST_CHECK_EQUAL(Solver(s, solutions), TxoutType::TAPROOT);
+    BOOST_CHECK_EQUAL(solutions.size(), 0U);
+
     // TxoutType::NULL_DATA
     s.clear();
     s << OP_RETURN << std::vector<uint8_t>({0}) << std::vector<uint8_t>({75})
@@ -175,6 +187,20 @@ BOOST_AUTO_TEST_CASE(script_standard_Solver_success) {
         s << OP_CHECKMULTISIG;
         BOOST_CHECK_EQUAL(Solver(s, solutions), TxoutType::NONSTANDARD);
         BOOST_CHECK_EQUAL(solutions.size(), 0);
+
+        // mutated TxoutType::TAPROOT -- commitment
+        s.clear();
+        s << OP_SCRIPTTYPE << OP_1;
+        AppendPush(s, pushdataop, ToByteVector(pubkeys[0]));
+        BOOST_CHECK_EQUAL(Solver(s, solutions), TxoutType::NONSTANDARD);
+        BOOST_CHECK_EQUAL(solutions.size(), 0);
+
+        // mutated TxoutType::TAPROOT -- state
+        s.clear();
+        s << OP_SCRIPTTYPE << OP_1 << ToByteVector(pubkeys[0]);
+        AppendPush(s, pushdataop, std::vector<uint8_t>(32));
+        BOOST_CHECK_EQUAL(Solver(s, solutions), TxoutType::NONSTANDARD);
+        BOOST_CHECK_EQUAL(solutions.size(), 0);
     }
 
     // also try pushing the num_signatures and num_pubkeys using PUSH_N opcode
@@ -260,6 +286,17 @@ BOOST_AUTO_TEST_CASE(script_standard_Solver_failure) {
     s.clear();
     s << OP_0 << std::vector<uint8_t>(19, 0x01);
     BOOST_CHECK_EQUAL(Solver(s, solutions), TxoutType::NONSTANDARD);
+
+    // TxoutType::TAPROOT with incorrectly sized commitment
+    s.clear();
+    s << OP_SCRIPTTYPE << OP_1 << std::vector<uint8_t>(30, 0x01);
+    BOOST_CHECK_EQUAL(Solver(s, solutions), TxoutType::NONSTANDARD);
+
+    // TxoutType::TAPROOT with incorrectly sized state
+    s.clear();
+    s << OP_SCRIPTTYPE << OP_1 << ToByteVector(pubkey)
+      << std::vector<uint8_t>(31, 0x01);
+    BOOST_CHECK_EQUAL(Solver(s, solutions), TxoutType::NONSTANDARD);
 }
 
 BOOST_AUTO_TEST_CASE(script_standard_ExtractDestination) {
@@ -299,6 +336,11 @@ BOOST_AUTO_TEST_CASE(script_standard_ExtractDestination) {
     s << OP_1 << ToByteVector(pubkey) << OP_1 << OP_CHECKMULTISIG;
     BOOST_CHECK(!ExtractDestination(s, address));
 
+    // TxoutType::TAPROOT
+    s.clear();
+    s << OP_SCRIPTTYPE << OP_1 << ToByteVector(pubkey);
+    BOOST_CHECK(!ExtractDestination(s, address));
+
     // TxoutType::NULL_DATA
     s.clear();
     s << OP_RETURN << std::vector<uint8_t>({75});
@@ -375,6 +417,11 @@ BOOST_AUTO_TEST_CASE(script_standard_ExtractDestinations) {
     BOOST_CHECK(boost::get<PKHash>(&addresses[1]) &&
                 *boost::get<PKHash>(&addresses[1]) == PKHash(pubkeys[1]));
 
+    // TxoutType::TAPROOT
+    s.clear();
+    s << OP_SCRIPTTYPE << OP_1 << ToByteVector(pubkeys[0]);
+    BOOST_CHECK(!ExtractDestinations(s, whichType, addresses, nRequired));
+
     // TxoutType::NULL_DATA
     s.clear();
     s << OP_RETURN << std::vector<uint8_t>({75});
diff --git a/src/test/transaction_tests.cpp b/src/test/transaction_tests.cpp
index 6f1f0874f..e3d5be2f5 100644
--- a/src/test/transaction_tests.cpp
+++ b/src/test/transaction_tests.cpp
@@ -825,6 +825,40 @@ BOOST_AUTO_TEST_CASE(test_IsStandard) {
     BOOST_CHECK(!IsStandardTx(CTransaction(t), reason));
     BOOST_CHECK_EQUAL(reason, "multi-op-return");
 
+    // Taproot
+    t.vout.resize(1);
+    t.vout[0].nValue = MAX_MONEY;
+    t.vout[0].scriptPubKey = CScript()
+                             << OP_SCRIPTTYPE << OP_1
+                             << ParseHex("02c192b7e5fd2c86d83d30ac8dd63eaf00c2f"
+                                         "ab891fda5ee885931e73d13c52452");
+    reason.clear();
+    BOOST_CHECK(IsStandardTx(CTransaction(t), reason));
+    t.vout[0].scriptPubKey = CScript()
+                             << OP_SCRIPTTYPE << OP_1
+                             << ParseHex("02c192b7e5fd2c86d83d30ac8dd63eaf00c2f"
+                                         "ab891fda5ee885931e73d13c524");
+    reason.clear();
+    BOOST_CHECK(!IsStandardTx(CTransaction(t), reason));
+    BOOST_CHECK_EQUAL(reason, "scriptpubkey");
+    t.vout[0].scriptPubKey = CScript()
+                             << OP_SCRIPTTYPE << OP_1
+                             << ParseHex("02c192b7e5fd2c86d83d30ac8dd63eaf00c2f"
+                                         "ab891fda5ee885931e73d13c52452")
+                             << ParseHex("858f2481a6c4df431e2fbdc1a27bc8a952fef"
+                                         "035af0c75d29ee62e9a5e46ab3e");
+    reason.clear();
+    BOOST_CHECK(IsStandardTx(CTransaction(t), reason));
+    t.vout[0].scriptPubKey = CScript()
+                             << OP_SCRIPTTYPE << OP_1
+                             << ParseHex("02c192b7e5fd2c86d83d30ac8dd63eaf00c2f"
+                                         "ab891fda5ee885931e73d13c52452")
+                             << ParseHex("858f2481a6c4df431e2fbdc1a27bc8a952fef"
+                                         "035af0c75d29ee62e9a5e46ab");
+    reason.clear();
+    BOOST_CHECK(!IsStandardTx(CTransaction(t), reason));
+    BOOST_CHECK_EQUAL(reason, "scriptpubkey");
+
     // Check large scriptSig (non-standard if size is >1650 bytes)
     t.vout.resize(1);
     t.vout[0].nValue = MAX_MONEY;
diff --git a/src/wallet/rpcdump.cpp b/src/wallet/rpcdump.cpp
index 298476490..cda1b76de 100644
--- a/src/wallet/rpcdump.cpp
+++ b/src/wallet/rpcdump.cpp
@@ -1156,6 +1156,7 @@ static std::string RecurseImportData(const CScript &script,
             return "";
         }
         case TxoutType::NULL_DATA:
+        case TxoutType::TAPROOT:
             return "unspendable script";
         case TxoutType::NONSTANDARD:
         default:
diff --git a/src/wallet/scriptpubkeyman.cpp b/src/wallet/scriptpubkeyman.cpp
index d1b384d00..170ded88a 100644
--- a/src/wallet/scriptpubkeyman.cpp
+++ b/src/wallet/scriptpubkeyman.cpp
@@ -96,6 +96,7 @@ IsMineResult IsMineInner(const LegacyScriptPubKeyMan &keystore,
     switch (whichType) {
         case TxoutType::NONSTANDARD:
         case TxoutType::NULL_DATA:
+        case TxoutType::TAPROOT:
             break;
         case TxoutType::PUBKEY:
             keyID = CPubKey(vSolutions[0]).GetID();
diff --git a/test/functional/logos_feature_taproot_key_spend.py b/test/functional/logos_feature_taproot_key_spend.py
index b4172ca9d..2838ec56d 100644
--- a/test/functional/logos_feature_taproot_key_spend.py
+++ b/test/functional/logos_feature_taproot_key_spend.py
@@ -139,8 +139,6 @@ class TaprootKeySpendTest(BitcoinTestFramework):
         self.num_nodes = 1
         self.setup_clean_chain = True
         self.extra_args = [['-whitelist=noban@127.0.0.1',
-                            '-allownonstdtxnconsensus=1',
-                            '-acceptnonstdtxn=1',
                             f'-replayprotectionactivationtime={ACTIVATION_TIME}']]
 
     def run_test(self):
diff --git a/test/functional/logos_feature_taproot_script_path.py b/test/functional/logos_feature_taproot_script_path.py
index 747aff79a..c6f364720 100644
--- a/test/functional/logos_feature_taproot_script_path.py
+++ b/test/functional/logos_feature_taproot_script_path.py
@@ -396,8 +396,6 @@ class TaprootScriptSpendTest(BitcoinTestFramework):
         self.num_nodes = 1
         self.setup_clean_chain = True
         self.extra_args = [['-whitelist=noban@127.0.0.1',
-                            '-allownonstdtxnconsensus=1',
-                            '-acceptnonstdtxn=1',
                             f'-replayprotectionactivationtime={ACTIVATION_TIME}']]
 
     def run_test(self):
-- 
2.39.5 (Apple Git-154)

